File: cpp/src/day02.cpp
#include <algorithm>
#include <cassert>
#include <fstream>
#include <iostream>
#include <numeric>
#include <optional>
#include <ranges>
#include <string>
#include <vector>

struct Set
{
	uint64_t red = 0;
	uint64_t green = 0;
	uint64_t blue = 0;

	bool isPossible(const Set& maxSet) const
	{
		return red <= maxSet.red
			&& green <= maxSet.green
			&& blue <= maxSet.blue;
	}

	uint64_t getPower() const { return red * green * blue; }
};

struct Game
{
	uint64_t id = 0;
	std::vector<Set> sets;

	Set getMinPossibleSet() const
	{
		return std::accumulate(begin(sets), end(sets), Set(), [] (Set minSet, const Set& set) -> Set
		{
			return
			{
				std::max(minSet.red, set.red),
				std::max(minSet.green, set.green),
				std::max(minSet.blue, set.blue),
			};
		});
	}
};

static Game parseGame(const std::string& gameString)
{
	Game game;
	game.sets.push_back({});

	uint64_t currentNum = 0;

	for (const auto word : std::views::split(gameString, ' '))
	{
		const std::string token(&*word.begin(), std::ranges::distance(word));

		Set& set = game.sets.back();

		if (token == "Game")
			continue;
		else if (game.id == 0)
			game.id = std::stoull(token);
		else if (std::isdigit(token.front()))
			currentNum = std::stoull(token);
		else if (token.find("red") != std::string::npos)
			set.red = currentNum;
		else if (token.find("green") != std::string::npos)
			set.green = currentNum;
		else if (token.find("blue") != std::string::npos)
			set.blue = currentNum;
		else
			assert(false);

		if (token.back() == ';')
			game.sets.push_back({});
	}

	return game;
}

static std::vector<Game> loadGames(const char* filename)
{
	std::vector<Game> games;

	std::fstream s{ filename, s.in };
	assert(s.is_open());

	while (s.peek() != EOF)
	{
		std::string line;
		std::getline(s, line);
		assert(line.size() > 0);

		games.push_back(parseGame(line));
		assert(games.back().id == games.size());
	}

	return games;
}

static uint64_t partOne(const std::vector<Game>& games)
{
	static constexpr Set maxSet = { 12, 13, 14 };
	static constexpr auto isPossible = [] (const Set& set) { return set.isPossible(maxSet); };

	return std::accumulate(begin(games), end(games), 0ull, [] (uint64_t sum, const Game& game)
	{
		return sum + (game.id * std::all_of(begin(game.sets), end(game.sets), isPossible));
	});
}

static uint64_t partTwo(const std::vector<Game>& games)
{
	return std::accumulate(begin(games), end(games), 0ull, [] (uint64_t sum, const Game& game)
	{
		return sum + game.getMinPossibleSet().getPower();
	});
}

static auto process(const char* filename)
{
	const std::vector<Game> games = loadGames(filename);
	return std::make_pair(partOne(games), partTwo(games));
}

static void processPrintAndAssert(const char* filename, std::optional<std::pair<uint64_t, uint64_t>> expected = {})
{
	const auto result = process(filename);
	std::cout << "Part 1: " << result.first << " Part 2: " << result.second << std::endl;
	assert(!expected || result == expected);
}

void day02()
{
	processPrintAndAssert("../data/02/test.txt", std::make_pair(8ull, 2286ull));
	processPrintAndAssert("../data/02/real.txt", std::make_pair(2447ull, 56322ull));
}

File: ts/src/day02.ts
import * as fs from "fs";
import { assert } from "./assert"

class Set {
	red: number;
	green: number;
	blue: number;

    constructor(r: number, g: number, b: number) {
        this.red = r;
        this.green = g;
        this.blue = b;
    }

    isPossible(maxSet: Set): boolean {
		return this.red <= maxSet.red
			&& this.green <= maxSet.green
			&& this.blue <= maxSet.blue;
	}

	getPower(): number {
        return this.red * this.green * this.blue;
    }
};

class Game {
	id: number = 0;
	sets: Set[] = [];

	getMinPossibleSet(): Set {
        let set: Set = new Set(0, 0, 0);

        this.sets.forEach((s:Set) => {
            set.red = Math.max(set.red, s.red);
            set.green = Math.max(set.green, s.green);
            set.blue = Math.max(set.blue, s.blue);
        });

        return set;
	}
};

function parseGame(gameString: string): Game {
	let game = new Game();
    assert(game.id === 0);
	game.sets.push(new Set(0, 0, 0));

	let currentNum: number = 0;

	gameString.split(' ').forEach((token: string) => {
		let set = game.sets.at(game.sets.length - 1);

		if (token == "Game")
			return;
		else if (game.id === 0)
			game.id = parseInt(token);
		else if ('0' <= token[0] && token[0] <= '9')
			currentNum = parseInt(token);
		else if (token.includes("red"))
			set.red = currentNum;
		else if (token.includes("green"))
			set.green = currentNum;
		else if (token.includes("blue"))
			set.blue = currentNum;
		else
			assert(false);

		if (token[token.length - 1] == ';')
            game.sets.push(new Set(0, 0, 0));
	});

	return game;
}

function loadGames(filename: string): Array<Game>
{
	let games: Game[] = [];

    const data: string = fs.readFileSync(filename, "utf-8");
    const lines: string[] = data.split('\n');

    lines.forEach((line: string) => {
        let game: Game = parseGame(line);
        games.push(game);
        assert(game.id == games.length, "id: " + game.id + ", length: " + games.length);
    });

	return games;
}

function partOne(games: Game[]): number
{
    const maxSet: Set = new Set(12, 13, 14);

    return games.reduce<number>((sum: number, game: Game) => {
        return game.sets.every((set: Set) => { return set.isPossible(maxSet)})
            ? sum + game.id
            : sum;
    }, 0);
}

function partTwo(games: Game[]): number
{
    return games.reduce<number>((sum: number, game: Game) => {
		return sum + game.getMinPossibleSet().getPower();
	}, 0);
}

function process(filename: string) {
	const games: Array<Game> = loadGames(filename);
	return { first: partOne(games), second: partTwo(games) };
}

function processPrintAndAssert(filename: string, expected = { first: 0, second: 0 })
{
	const result = process(filename);
    console.log("Part 1: " + result.first + " Part 2: " + result.second);
	assert(!expected || expected.first === 0 || result.first === expected.first);
	assert(!expected || expected.second === 0 || result.second === expected.second);
}

export function day02() {
	processPrintAndAssert("../data/02/test.txt", { first: 8, second: 2286 });
	processPrintAndAssert("../data/02/real.txt", { first: 2447, second: 56322 });
}
